annotation issues

If we enable securesystemslib type checks (which we can't really yet do because https://github.com/secure-systems-lab/securesystemslib/issues/944), it shows two issues in python-tuf:


```
tuf/api/serialization/json.py:101: error: Item "None" of "str | None" has no attribute "encode"  [union-attr]
```

This looks trivial to fix: `encode_canonical()` can return None but only if it's given the output_function argument. We can just do `assert canonical_str is not None` to silence this

```
tuf/api/dsse.py:84: error: Incompatible return value type (got "Envelope", expected "SimpleEnvelope[Root]")  [return-value]
tuf/api/dsse.py:84: error: Incompatible return value type (got "Envelope", expected "SimpleEnvelope[Timestamp]")  [return-value]
tuf/api/dsse.py:84: error: Incompatible return value type (got "Envelope", expected "SimpleEnvelope[Snapshot]")  [return-value]
tuf/api/dsse.py:84: error: Incompatible return value type (got "Envelope", expected "SimpleEnvelope[Targets]")  [return-value]
```

not entirely sure how to fix this one: maybe just `typing.cast()` call?

--- a/tuf/api/dsse.py
+++ b/tuf/api/dsse.py
@@ -81,7 +81,7 @@ class SimpleEnvelope(Generic[T], BaseSimpleEnvelope):
         except Exception as e:
             raise DeserializationError from e
 
-        return envelope
+        return cast(SimpleEnvelope[T], envelope)
 
     def to_bytes(self) -> bytes:
         """Return envelope as JSON bytes.

--- a/tuf/api/serialization/json.py
+++ b/tuf/api/serialization/json.py
@@ -98,7 +98,10 @@ class CanonicalJSONSerializer(SignedSerializer):
         """
         try:
             signed_dict = signed_obj.to_dict()
-            canonical_bytes = encode_canonical(signed_dict).encode("utf-8")
+            canon_str = encode_canonical(signed_dict)
+            # encode_canonical cannot return None if output_function is not set
+            assert canon_str is not None  # noqa: S101
+            canonical_bytes = canon_str.encode("utf-8")
 
         except Exception as e:
             raise SerializationError from e

